### ISSUE TITLE
Apply expression can be omitted wrong suggestion

### ISSUE THREAD

**Issue by sqaaa** [detekt/detekt#2821]
I can't find issue relating to same problem as mine, so I going to text a new one.
There is a case like in screenshot
![image](https://user-images.githubusercontent.com/11176780/85551897-204d3080-b62b-11ea-9229-c6e312cb4c72.png)
Where I need to use ?.apply to save call not null instance of activity and then using it as a context to proceeding my logic
From my point of view there is no other case how this apply could be omitted without redo it in other way like this
![image](https://user-images.githubusercontent.com/11176780/85552300-889c1200-b62b-11ea-98db-a12e60b90282.png)
but it is not omission, isn't it?

* Version of detekt used: 1.9.1
* Version of Gradle used (if applicable): 3.6.3

**Comment by BraisGabin** [comment]
`let` is a better fit than `apply` for your use case. If `let` is marked as unnecessary we have a false positive. Could you check? 

Opinated answer: there's nothing wrong with using a local variable and an if. It's shorter that your second option and easier to read.

**Comment by sqaaa** [comment]
Yep! It's going to work with "let". Would you describe why do you think "let" is better than "apply" please?

**Comment by BraisGabin** [comment]
If I need to be clear here... I really don't know why. I know that when you just use an scoped function to check the nullabiblity the "convention" is `let`.

I use `apply` for initialize classes that have a bad API (because they are not initialized just with the constructor). For example:

```kotlin
val redPaint = Paint().apply {
  setColor(RED)
}
```

You could do the same with `let`

```kotlin
val redPaint = Paint().let {
  it.setColor(RED)
  it
}
```

but as you can see is not that "nice"

And you could even use `also`:

```kotlin
val redPaint = Paint().also {
  it.setColor(RED)
}
```

But they are not the correct fit here. I'm sorry but I can't give a rule here. If you search for "scope functions kotlin" in your favorite search engine you will find plenty of documentation about this topic.

I'm going to keep this open waiting for the feedback of other contributors because I have a concern here: Should we point out that you should use `let` instead? Can we decide this statically?

**Comment by cortinico** [comment]
> Yep! It's going to work with "let". Would you describe why do you think "let" is better than "apply" please?

As a rule of thumb, you can use the list you can find here: https://kotlinlang.org/docs/reference/scope-functions.html#function-selection

Your case is: "Executing a lambda on non-null objects" -> `let`

This rule is to prevent abuse of the `apply{}` scope function to configure just one field of an object. The example that @BraisGabin pointed out is the actual edge case:

```kotlin
val redPaint = Paint().apply {
  setColor(RED)
}
```

I'll argue this could potentially be rewritten with:

```kotlin
val redPaint = Paint()
redPaint.setColor(RED)
```

> I'm going to keep this open waiting for the feedback of other contributors because I have a concern here: Should we point out that you should use let instead? Can we decide this statically?

As for reporting an alternative, I don't think it's easier to compute what's the valid alternative. Also the rationale behind the rule should be: if you use `apply{}` only to update one field -> probably you don't need apply at all.

**Comment by schalkms** [comment]
Should the documentation of this rule be improved?

**Comment by BraisGabin** [comment]
After reading the code I think that we can improve the documentation AND the message:

- We can add one more example of `noncompliant` pointing that this: `config?.apply { println(version) }` should be `config?.let { println(it.version) }`
- And we can check if the user is calling the apply using the safe null operator `?.`. In this case we can tell that he should use `let` instead of `apply`. I mean, it's not obvious why `config?.let { println(it.version) }` is better than `config?.apply { println(version) }`, so if we can point it out we will help a lot.

What do you think?

**Comment by schalkms** [comment]
@BraisGabin good ideas!

**Comment by BraisGabin** [comment]
Ok! Now I know why in `config?.apply { println(version) }` `let` is recommended. With `apply` you can't be sure which version you are called. The version from config or the version from the outer class? With let this is more obvious because you need to specify `it` if you want to access the `version` from config.

So, it's not just a convention. There's an actual reason.
